lnd_test.go

```diff
@@ -2036,6 +2036,134 @@ func testChannelBalance(net *lntest.NetworkHarness, t *harnessTest) {
+		lntest.OpenChannelParams{
+			Amt: chanAmt,
+		},
+	)
+
+	// Wait for Alice and Carol to receive the channel edge from the
+	// funding manager.
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	err = net.Alice.WaitForNetworkChannelOpen(ctxt, chanPointAlice)
+	if err != nil {
+		t.Fatalf("alice didn't see the alice->carol channel before "+
+			"timeout: %v", err)
+	}
+
+	err = carol.WaitForNetworkChannelOpen(ctxt, chanPointAlice)
+	if err != nil {
+		t.Fatalf("alice didn't see the alice->carol channel before "+
+			"timeout: %v", err)
+	}
+
+	// Channel should be ready for payments.
+	const (
+		payAmt      = 100
+		numInvoices = 6
+	)
+
+	// Create a paystream from Alice to Carol to enable Alice to make
+	// a series of payments.
+	alicePayStream, err := net.Alice.SendPayment(ctxb)
+	if err != nil {
+		t.Fatalf("unable to create payment stream for alice: %v", err)
+	}
+
+	// Send payments from Alice to Carol a number of numInvoices
+	// times.
+	carolPubKey := carol.PubKey[:]
+	for i := 0; i < numInvoices; i++ {
+		err = alicePayStream.Send(&lnrpc.SendRequest{
+			Dest:           carolPubKey,
+			Amt:            int64(payAmt),
+			PaymentHash:    makeFakePayHash(t),
+			FinalCltvDelta: defaultBitcoinTimeLockDelta,
+		})
+		if err != nil {
+			t.Fatalf("unable to send alice htlc: %v", err)
+		}
+	}
+
+	// Test that the UnsettledBalance for both Alice and Carol
+	// is equal to the amount of invoices * payAmt.
+	var unsettledErr error
+	nodes := []*lntest.HarnessNode{net.Alice, carol}
+	err = lntest.WaitPredicate(func() bool {
+		// There should be a number of PendingHtlcs equal
+		// to the amount of Invoices sent.
+		unsettledErr = assertNumActiveHtlcs(nodes, numInvoices)
+		if unsettledErr != nil {
+			return false
+		}
+
+		// Set the amount expected for the Unsettled Balance for
+		// this channel.
+		expectedBalance := numInvoices * payAmt
+
+		// Closure helper to check Unsettled Balance.
+		assertUnsettledBalance := func(expected int,
+			nodes ...*lntest.HarnessNode) error {
+
+			// Check each nodes UnsettledBalance field.
+			ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+			for _, node := range nodes {
+				// Get channel info for the node.
+				chanInfo, err := getChanInfo(ctxt, node)
+				if err != nil {
+					return err
+				}
+
+				// Check that UnsettledBalance is what we expect.
+				if int(chanInfo.UnsettledBalance) != expected {
+					return fmt.Errorf("unsettled balance failed "+
+						"expected: %v, received: %v", expected,
+						chanInfo.UnsettledBalance)
+				}
+			}
+			return nil
+		}
+
+		unsettledErr = assertUnsettledBalance(expectedBalance, carol, net.Alice)
+		if unsettledErr != nil {
+			return false
+		}
+
+		return true
+	}, time.Second*15)
+	if err != nil {
+		t.Fatalf("unsettled balace error: %v", unsettledErr)
+	}
+
+	// Force and assert the channel closure.
+	ctxt, _ = context.WithTimeout(ctxb, channelCloseTimeout)
+	closeChannelAndAssert(ctxt, t, net, net.Alice, chanPointAlice, true)
```

Comment:
We should also use `cleanupForceClose` when force closing a channel to mine the sweeping transactions and not pollute the other tests with them, as it can break certain mempool assumptions within them.

Reply:
fixed